python/lsst/scarlet/lite/operators_pybind11.cc

```diff
@@ -82,65 +83,65 @@ void apply_filter(
 // and create a boolean map of "orphans" that are non-monotonic in all directions.
 template <typename M>
 void get_valid_monotonic_pixels(
-    const int i,
-    const int j,
+    const int start_i,
+    const int start_j,
     Eigen::Ref<M, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> image,
     Eigen::Ref<MatrixB, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> unchecked,
     Eigen::Ref<MatrixB, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> orphans,
     const double variance,
     Eigen::Ref<Bounds, 0, Eigen::Stride<4, 1>> bounds,
     const double thresh=0
-){
-    // Check the pixel below this one
-    if(i>0 && unchecked(i-1,j)){
-        if(image(i-1,j) <= image(i,j)+variance && image(i-1,j) > thresh){
-            unchecked(i-1,j) = false;
-            orphans(i-1,j) = false;
-            if(i-1 < bounds(0)){
-                bounds(0) = i-1;
-            }
-            get_valid_monotonic_pixels(i-1, j, image, unchecked, orphans, variance, bounds);
-        } else {
-            orphans(i-1,j) = true;
-        }
-    }
-    // Check the pixel above this one
-    if(i < image.rows()-1 && unchecked(i+1,j)){
-        if(image(i+1,j) <= image(i,j)+variance && image(i+1,j) > thresh){
-            unchecked(i+1,j) = false;
-            orphans(i+1,j) = false;
-            if(i+1 > bounds(1)){
-                bounds(1) = i+1;
+) {
+    // Create a queue to store pixels that need to be processed
+    std::queue<std::pair<int, int>> pixel_queue;
+    pixel_queue.push({start_i, start_j});
+
+    while (!pixel_queue.empty()) {
+        auto [i, j] = pixel_queue.front();
+        pixel_queue.pop();
+
+        // Define directions: down, up, left, right
+        const int di[] = {-1, 1, 0, 0};
+        const int dj[] = {0, 0, -1, 1};
+        const int bounds_idx[] = {0, 1, 2, 3}; // corresponding bounds indices
+
+        // Check all four directions
+        for (int dir = 0; dir < 4; dir++) {
+            int ni = i + di[dir];
+            int nj = j + dj[dir];
+
+            // Check bounds
+            if (ni < 0 || ni >= image.rows() || nj < 0 || nj >= image.cols()) {
```

Comment:
I'd have `n_rows = image.rows()` outside of both of these loops rather than hoping that the compiler will optimize it away (which I'm not sure it can here).

Also suggest `!((ni >= 0) && (ni < n_rows) && (nj >= 0 ) &&  (nj < n_cols))` for a couple of reasons:
- brackets make multiple comparisons clearer (IMO) and avoid potential operator precedence mishaps
- checking if comparisons are true is nan-safe (even if there shouldn't ever be a nan here)

Reply:
A couple of points:
- image.rows() and image.cols() are only used once each, so I don't see the need to add extra variables.
- I personally don't like extra brackets when the operator precedence is known but I know that in numpy the above line would be misinterpreted so did make this change just to make the intention clear.
- Good catch about nan-safety in general, although these are integers so that shouldn't matter and I don't think that it's necessary here. But I did also insert extra parentheses for you.